File: fixtures/scale/dart/medium/lib/lexer/char_stream.dart
class CharStream {
  final String source;
  int _index = 0;

  CharStream(this.source);

  bool get hasNext => _index < source.length;

  String peek() => source[_index];

  String next() => source[_index++];

  int get position => _index;

  void skipWhitespace() {
    while (hasNext && peek().trim().isEmpty) _index++;
  }

  String takeWhile(bool Function(String) accept) {
    final sb = StringBuffer();
    while (hasNext && accept(peek())) sb.write(next());
    return sb.toString();
  }
}

File: fixtures/scale/java/medium/src/calc/lexer/CharStream.java
package calc.lexer;

import java.util.function.IntPredicate;

public class CharStream {
    private final String source;
    private int index;

    public CharStream(String source) {
        this.source = source;
    }

    public boolean hasNext() { return index < source.length(); }

    public char peek() { return source.charAt(index); }

    public char next() { return source.charAt(index++); }

    public int position() { return index; }

    public void skipWhitespace() {
        while (hasNext() && Character.isWhitespace(peek())) {
            index++;
        }
    }

    public String takeWhile(IntPredicate accept) {
        StringBuilder sb = new StringBuilder();
        while (hasNext() && accept.test(peek())) {
            sb.append(next());
        }
        return sb.toString();
    }
}
